### PR TITLE
RSDK-5269 - CLI Support single file uploads and auto-compression

### PR DESCRIPTION
**Pull request by zaporter-work**
Ticket: https://viam.atlassian.net/browse/RSDK-5269

This adds the ability to provide a path to a single file or a path to a directory to the `viam module upload` subcommand. Either of these will be automatically compressed and added to a temporary tarball that will be uploaded instead.

There is a bit of optimization left on the table because the validate step scans through the archive (after it has been written) but the additional code complexity to take advantage of this inefficiency did not seem worth the cost. (I tested. The validation step was negligible) 

I added the ability to include a whole directory because this is a common use case for cpp modules to ship separate libraries that are still required (so including the entire install dir is reasonable and we dont need to make people tar it up before they ship it) (open to debate if we want to support this)

**Open question**
Do we want to support uploading a list of files ex:
`viam module upload --platform xxx --version yyy ./src requirements.txt run.sh`
This could eliminate the need for the tar step (which is confusing). However, it also comes at the cost that now people will need to remember what files are required in their module. Fahmina was not opposed but I keep going back and forth.

### CODE REVIEW

File: cli/app.go
Comment (zaporter-work) [inline cli/app.go:840]: cleanup

File: cli/module_registry.go
Comment (zaporter-work) [inline cli/module_registry.go:651]: This is a very weak check. I am on the fence as to whether we should make it a better test using a tar + gzip reader but that would require an extra scan. Opinions?

File: cli/module_registry.go
Comment (zaporter-work) [inline cli/module_registry.go:246]: Defer UncheckedErrorFunc wasn't much cleaner. Open to suggestions.

File: cli/boards.go
Comment (zaporter-work) [inline cli/boards.go:326]: Tidiness. This came up for another app ticket. Without this check, we attempt to untar the http error message and fail.

File: cli/module_registry.go
Comment (zaporter-work) [inline cli/module_registry.go:654]: We want to create a file instead of using a byte buffer (like board) because a module can be very large (multi GiB).

File: cli/module_registry.go
Comment (zaporter-work) [inline cli/module_registry.go:652]: !strings.HasSuffix(strings.ToLower(file.Name()), ".tar.gz") &&
		!strings.HasSuffix(strings.ToLower(file.Name()), ".tgz")

File: cli/archive_test.go
Comment (benjirewis) [inline cli/archive_test.go:36]:
```suggestion
			t.Errorf("file %s was not found", f)
```
[nit] test errors should still be lowercase.

File: cli/module_registry.go
Comment (benjirewis) [inline cli/module_registry.go:187]:
```suggestion
		return errors.New("no package to upload -- please provide a path to your module. Use --help for more information")
```

File: cli/module_registry.go
Comment (benjirewis) [reply cli/module_registry.go:246]: If you aren't attached to logging an error from inability to remove, you could use [`utils.RemoveFileNoError`](https://github.com/viamrobotics/rdk/blob/main/utils/file.go#L29). We use that to remove generated `.sock` files in module management code. It also only removes the file if it exists.

**Review by benjirewis — COMMENTED** [review]
Super cool! I don't have a ton of input on the "usefulness" of specifying a single file vs. multiple files vs. a whole directory, so I'll leave that to others.

File: cli/module_registry.go
Comment (benjirewis) [reply cli/module_registry.go:652]: > to validate integrity

What could you validate about the created tarball? That it was `tar`able?

File: cli/boards.go
Comment (benjirewis) [inline cli/boards.go:334]: Is this limit no longer relevant (I think you mentioned this in another comment but wanted to verify)?

File: cli/module_registry.go
Comment (benjirewis) [reply cli/module_registry.go:652]: Also why are you checking `!strings.HasSuffix(strings.ToLower(path), ".tgz")`? Won't this return `true` for most files?

File: cli/archive.go
Comment (abe-winter) [inline cli/archive.go:28]: why assemble a map + convert to slice? deduping?

File: cli/archive.go
Comment (abe-winter) [inline cli/archive.go:27]: might want to do one or both of:
- log here that you're excluding irregular file and say the path
- note in docs that symlinks and other special files are excluded

I have gotten bitten before by systems that ignore symlinks

File: cli/archive.go
Comment (abe-winter) [inline cli/archive.go:48]: I don't know what this does exactly but FYI docs say 'never use this for closing writers'

https://github.com/viamrobotics/goutils/blob/23898314dc98153baf6f45dd3ec142cd92200639/error.go#L69

File: cli/archive.go
Comment (abe-winter) [reply cli/archive.go:70]: +1 for logging once per file -- the upload logs are too chatty in CI (where `\r` doesn't work)

File: cli/boards.go
Comment (abe-winter) [inline cli/boards.go:338]: does the header.name override on archive.go L100 cause slightly different behavior here? (full path vs basename?). does it matter?

File: cli/module_registry.go
Comment (abe-winter) [inline cli/module_registry.go:187]: this isn't always a 'package' anymore, is it?

File: cli/module_registry.go
Comment (abe-winter) [inline cli/module_registry.go:454]: 'provided' might no longer be right here if the archive was created

**Review by abe-winter — APPROVED** [review]
left a bunch of nits but nothing obviously wrong

File: cli/archive.go
Comment (zaporter-work) [reply cli/archive.go:28]: Yup, deduping. This isn't the prettiest pattern but it is used in lots of places throughout our codebase. If there is a better pattern happy to swap to that.

**Comment by michaellee1019** [comment]
> Open question
> Do we want to support uploading a list of files ex:

Im torn too:

I think no, people can run the tar command if they need this. The main purpose is to remove a step for those that have a single file. I like the directory idea you added so lets keep.

I think yes because we'll likely need it in the CI v2 project. Developers will likely need to tell us what files to include.

How about we wait until CI v2 scope to decide this?

**Review by michaellee1019 — APPROVED** [review]
Already got a ton of feedback and review from Abe and Benji. I reviewed this functionally (how the commands work) and it LGTM!

**Review by benjirewis — APPROVED** [review]
Nice! LGTM 🧑‍🔧 .

**Pull request merged**
